### ISSUE TITLE
Stream audio chunk by chunk to Whisper

### ISSUE THREAD

**Issue by mat-hek** [elixir-nx/bumblebee#261]
Hey, it's already possible to provide data to the model in chunks, however, it seems that the whole audio has to be available at once, which is impossible for live streaming. Would it be possible to support streaming the audio to the serving chunk by chunk?

**Comment by josevalim** [comment]
We do support streaming in latest Bumblebee but we only support streaming files. We would need to improve the API so you can pass your own stream to Whisper and then we transform it. :) So it is not possible yet but we may be 90% there.

**Comment by mat-hek** [comment]
Hmm, correct me if I'm wrong, but it seems you load the entire file to the memory upfront [here](https://github.com/elixir-nx/bumblebee/blob/main/lib/bumblebee/audio/speech_to_text_whisper.ex#L586) :D

**Comment by josevalim** [comment]
We do, but that could be worked out by doing multiple ffmpeg calls. My point is that the complexity now is in the stream composition/processing, not in Nx/Axon/etc. And the former is much easier!

**Comment by mat-hek** [comment]
So it's just about rewriting `client_preprocessing` [here](https://github.com/elixir-nx/bumblebee/blob/main/lib/bumblebee/audio/speech_to_text_whisper.ex#L76) from Enum to Stream? How would batching be handled then?

**Comment by josevalim** [comment]
The serving does the batching, although ideally you want to chunk the stream to match the server batch size too.

**Comment by jonatanklosko** [comment]
Yeah, reading chunks separately from disk is on my list, but it's just an optimisation so we released without it.

As for accepting a stream as serving input, it's a bit different, but definitely doable. Note that for large audio when we split into multiple chunks, we make the chunks overlap. So for a 50s audio we would transcribe like 0-30 and 20-50, this way each transcription has some context and we merge the overlaps accordingly. So if we are given an stream, we need to accumulate until the right size and emit overlapping chunks.

**Comment by mat-hek** [comment]
> this way each transcription has some context and we merge the overlaps accordingly

Yeah, that's actually the reason I'd like to stream to the serving instead of running it for each chunk (as I do now). To make it 'live', I'd need to have at most a few seconds long chunks, but from the docs I see that the default is 5 seconds ;)

**Comment by jonatanklosko** [comment]
> hmm, I don't know where I found these 5 seconds, it seems it's just what I set

The default context is 1/6 of the chunk length, for whisper the chunk is 30s, so the context is 5s (both sides, so it's a 10s overlap).

I'm not sure if we can reasonably handle arbitrarily small chunks (especially as we do context, because then the context would be very small). So I would imagine we accumulate first 30s, then next 20s, next 20s.

**Comment by jonatanklosko** [comment]
> I would imagine we accumulate first 30s, then next 20s, next 20s.

Ah, we should accumulate whatever is the `chunk_length`, so yeah it could as well be smaller.

**Comment by josevalim** [comment]
Yeah, we can probably transform the stream to either split or accumulate batch size. We can also just do nothing and tell the user that whatever audio size they pass, it will be sent as is, so the buffering is on their own. The latter is the most flexible and likely the simpler too.

**Comment by mat-hek** [comment]
> If we stream, we will likely expect pcm chunks

Seems very reasonable

> conversion would be up to you

You can use Membrane for that too 😄 here's a PR with a Livebook example: https://github.com/membraneframework/membrane_demo/pull/249

**Comment by lawik** [comment]
Is there a difference between it accepting a real stream and repeatedly calling it with the chunk size you want processed?

I guess the current functionality for improving the edges of chunks with overlap and so on suffer when I just send it exactly sized chunks?

@mat-hek as you would know it is not particularly hard to get an appropriate slice of PCM to send it out of Membrane :D.

**Comment by jonatanklosko** [comment]
@lawik the idea is that we get a stream of continuous chunks, but we would still do overlapping as part of preprocessing and then merging in postprocessing to improve the output.

**Comment by lawik** [comment]
Awesome!

**Comment by linusdm** [comment]
Is this discussion targetted at enabling Whisper specifically? Or will these improvements also allow other more general audio processing models (e.g. [audio classification models](https://huggingface.co/models?pipeline_tag=audio-classification&sort=trending)) to benefit from this streaming solution?

**Comment by jonatanklosko** [comment]
#361 enables input streaming.

Thinking more about this, I'm not entirely sure if the context overlapping algorithm is very going to be effective with small chunks (as needed for live transcription). The way the algorithm works is that we transcribe two subsequent overlapping chunks of audio, and they should result in two sentences overlapping *to some extent* at the edges. Then we merge the overlaps to hopefully get the right transcription from the left chunk and from the right chunk. The issue with small chunks is that the sentences are short and there may be very few if any overlapping words.

So for short chunks it may be better to not use the chunking and have some other logic, such as splitting input at low amplitude points to avoid cutting mid-word.

These are just high-level thoughts though!

**Comment by samrat** [comment]
Hello,

I'm trying to use this in a Livebook using kino_live_audio: https://gist.github.com/samrat/fc5792bfc870ad887f29d4a944cafd7d . I'm passing a Stream to the serving, but I'm not seeing any output. Could you help me figure out what I'm doing wrong?

**Comment by jonatanklosko** [comment]
@samrat the main issue is that you are doing `Enum.map` instead of `Stream.map`, so it starts the stream at that point and blocks further execution :) Here's a more minimised example:

<details><summary>.livemd</summary>

````markdown
<!-- livebook:{"app_settings":{"access_type":"public","output_type":"rich","show_source":true,"slug":"vad"}} -->

# Streaming whisper

```elixir
Mix.install(
  [
    {:kino_live_audio, "~> 0.1"},
    {:nx, "~> 0.7.1"},
    {:bumblebee, github: "elixir-nx/bumblebee"},
    {:exla, ">= 0.0.0"},
    {:kino, github: "livebook-dev/kino", override: true}
  ],
  config: [nx: [default_backend: EXLA.Backend]]
)
```

## Section

```elixir
{:ok, model_info} = Bumblebee.load_model({:hf, "openai/whisper-tiny"})
{:ok, featurizer} = Bumblebee.load_featurizer({:hf, "openai/whisper-tiny"})
{:ok, tokenizer} = Bumblebee.load_tokenizer({:hf, "openai/whisper-tiny"})
{:ok, generation_config} = Bumblebee.load_generation_config({:hf, "openai/whisper-tiny"})

serving =
  Bumblebee.Audio.speech_to_text_whisper(
    model_info,
    featurizer,
    tokenizer,
    generation_config,
    compile: [batch_size: 1],
    chunk_num_seconds: 6,
    context_num_seconds: 2,
    stream: true,
    defn_options: [compiler: EXLA]
  )

Kino.start_child({Nx.Serving, serving: serving, name: WhisperServing})
```

```elixir
liveAudio = KinoLiveAudio.new(chunk_size: 1, unit: :s, sample_rate: featurizer.sampling_rate)
```

```elixir
audio_stream =
  liveAudio
  |> Kino.Control.stream()
  |> Stream.map(fn %{chunk: data} ->
    Nx.tensor(data, backend: Nx.BinaryBackend)
    |> Nx.stack()
    |> Nx.reshape({:auto, 1})
    |> Nx.mean(axes: [1])
  end)

frame = Kino.Frame.new() |> Kino.render()

for chunk <- Nx.Serving.batched_run(WhisperServing, audio_stream) do
  Kino.Frame.append(frame, Kino.Text.new(chunk.text, chunk: true))
end
```
````

</details> 

Sidenote: if you look at the console logs and the chunks are not being produced, it may be because the page was denied microphone access.
